### PR TITLE
Fix compute budget bump

### LINKED ISSUE #21225: Tx that succeeds with v1.8.2 fails with master
#### Problem
I recently encountered `bank hash frozen` mismatch while running a validator on testnet. This happened for a single slot, and resulted in my validator entering a loop of the following steps
a) Repair slot
b) Compute bank hashes
c) Check bank hash ==> see wrong value ==> purge ==> goto step a)

As such, the validator got stuck and eventually OOM'd.

With `v1.8.2`, I'd get:
```
[... INFO  solana_runtime::bank] bank frozen: 100608297
hash: Dp6V6h9Ky11u7msy9Et16Xd4GNVykUxFKCE76Jne7jX2
accounts_delta: 5ZG3koGMqBtNou8RFSeMmpUsU4UoMpUhhBuJWdSDbmRL
signature_count: 3953
last_blockhash: Bu2JiSgYJvyadPAFTef6Nh8RodWDbf2cobMSdbb3EDCE
capitalization: 1410898880597670717
...
[... INFO  solana_runtime::bank] accounts hash slot: 100608297 stats: BankHashStats {
num_updated_accounts: 8390,
num_removed_accounts: 0,
num_lamports_stored: 1087197652619615295,
total_data_len: 12606306,
num_executable_accounts: 0 }
```

With `master`, I'd get:
```
[... INFO  solana_runtime::bank] bank frozen: 100608297
hash: 6vyVwtcbhfLiM1skR4D7XT6X9ZHuXgxFL7sndu2QAd1
accounts_delta: 5jaEsNMBqF2pLzag7yfmv55MZA8kxwgRDGGmpcxUZf4Y
signature_count: 3953
last_blockhash: Bu2JiSgYJvyadPAFTef6Nh8RodWDbf2cobMSdbb3EDCE
capitalization: 1410898880597670717
...
[... INFO  solana_runtime::bank] accounts hash slot: 100608297 stats: BankHashStats {
num_updated_accounts: 8389,
num_removed_accounts: 0,
num_lamports_stored: 1087197651706463295,
total_data_len: 12475234,
num_executable_accounts: 0 }
```

The `blockhash` lines up so we know these two are operating on same set of shreds; however, the `accounts_delta` differs so there is a difference in transactions executing. This is further evidenced by the difference of 1 in `num_updated_accounts`: `8390 (v1.8.2)` vs. `8389 (master)`.

I dumped all of the updated accounts out, and found the account that wasn't updated by master was: `GK6qcLhoqW5GqfV1FuVihdr8cAW1xjUxn4kuQZssW1od`.
I checked this account vs `ledger-tool slot` output and found that this account was only touched by a single transaction (Entry 13, TX 0):
```
Transaction { signatures: [2wggxpkc28i9T4N339D3dHhUDweKDtNJyCjCLGx7ZnFJNa1G6iHcxDKMEoy3Rs1LJW6c973ataRKcX9KaGaninWB], message: Message { header: MessageHeader { num_required_signatures: 1, num_readonly_signed_accounts: 0, num_readonly_unsigned_accounts: 1 }, account_keys: [EJUKLLjBMhFnkonfn7wcThnHyDewmhVmG9sEuVP9cvF8, GK6qcLhoqW5GqfV1FuVihdr8cAW1xjUxn4kuQZssW1od, eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU], recent_blockhash: EpoMf68Y6cntiVKKBgB2f8dxWQodJUwpBTah9EmcQTq5, instructions: [CompiledInstruction { program_id_index: 2, accounts: [1, 0], data: [0, 0, 0, 0, 0, 0, 0, 0, 100, 2, 0, 0, 0, 0, 0, 0, 195, 243, 23, 62, 82, 33, 84, 138, 85, 104, 126, 40, 115, 34, 108, 200, 192, 9, 238, 0, 69, 140, 134, 56, 226, 149, 5, 239, 213, 29, 58, 213, 62, 179, 17, 64, 37, 171, 130, 163, 244, 213, 197, 198, 77, 221, 246, 2, 223, 103, 55, 169, 45, 164, 6, 78, 53, 35, 251, 202, 88, 201, 117, 114, 1, 27, 2, 0, 0, 0, 0, 0, 0, 249, 2, 24, 128, 132, 59, 154, 202, 0, 131, 3, 70, 203, 128, 128, 185, 2, 6, 96, 128, 96, 64, 82, 52, 128, 21, 97, 0, 16, 87, 96, 0, 128, 253, 91, 80, 96, 0, 128, 84, 96, 1, 96, 1, 96, 160, 27, 3, 25, 22, 51, 23, 144, 85, 97, 1, 212, 128, 97, 0, 50, 96, 0, 57, 96, 0, 243, 254, 96, 128, 96, 64, 82, 52, 128, 21, 97, 0, 16, 87, 96, 0, 128, 253, 91, 80, 96, 4, 54, 16, 97, 0, 76, 87, 96, 0, 53, 96, 224, 28, 128, 99, 9, 0, 240, 16, 20, 97, 0, 81, 87, 128, 99, 141, 165, 203, 91, 20, 97, 0, 102, 87, 128, 99, 251, 219, 173, 60, 20, 97, 0, 150, 87, 128, 99, 253, 172, 213, 118, 20, 97, 0, 173, 87, 91, 96, 0, 128, 253, 91, 97, 0, 100, 97, 0, 95, 54, 96, 4, 97, 1, 85, 86, 91, 97, 0, 192, 86, 91, 0, 91, 96, 0, 84, 97, 0, 121, 144, 96, 1, 96, 1, 96, 160, 27, 3, 22, 129, 86, 91, 96, 64, 81, 96, 1, 96, 1, 96, 160, 27, 3, 144, 145, 22, 129, 82, 96, 32, 1, 91, 96, 64, 81, 128, 145, 3, 144, 243, 91, 97, 0, 159, 96, 1, 84, 129, 86, 91, 96, 64, 81, 144, 129, 82, 96, 32, 1, 97, 0, 141, 86, 91, 97, 0, 100, 97, 0, 187, 54, 96, 4, 97, 1, 133, 86, 91, 97, 1, 61, 86, 91, 96, 0, 84, 96, 1, 96, 1, 96, 160, 27, 3, 22, 51, 20, 21, 97, 1, 58, 87, 96, 1, 84, 96, 64, 81, 99, 126, 214, 106, 187, 96, 225, 27, 129, 82, 130, 145, 96, 1, 96, 1, 96, 160, 27, 3, 131, 22, 145, 99, 253, 172, 213, 118, 145, 97, 1, 6, 145, 96, 4, 1, 144, 129, 82, 96, 32, 1, 144, 86, 91, 96, 0, 96, 64, 81, 128, 131, 3, 129, 96, 0, 135, 128, 59, 21, 128, 21, 97, 1, 32, 87, 96, 0, 128, 253, 91, 80, 90, 241, 21, 128, 21, 97, 1, 52, 87, 61, 96, 0, 128, 62, 61, 96, 0, 253, 91, 80, 80, 80, 80, 80, 91, 80, 86, 91, 96, 0, 84, 96, 1, 96, 1, 96, 160, 27, 3, 22, 51, 20, 21, 97, 1, 58, 87, 96, 1, 85, 86, 91, 96, 0, 96, 32, 130, 132, 3, 18, 21, 97, 1, 103, 87, 96, 0, 128, 253, 91, 129, 53, 96, 1, 96, 1, 96, 160, 27, 3, 129, 22, 129, 20, 97, 1, 126, 87, 96, 0, 128, 253, 91, 147, 146, 80, 80, 80, 86, 91, 96, 0, 96, 32, 130, 132, 3, 18, 21, 97, 1, 151, 87, 96, 0, 128, 253, 91, 80, 53, 145, 144, 80, 86, 254, 162, 100, 105, 112, 102, 115, 88, 34, 18, 32, 67, 229, 136, 143, 169, 218, 129, 82, 241, 34, 92, 20, 64, 223, 163, 31, 107, 4, 149, 242, 43, 128, 150, 60, 250, 178, 22, 180, 107, 138, 21, 139, 100, 115, 111, 108, 99, 67, 0, 8, 9, 0, 51, 111, 128, 128] }] } }
```
Replaying the slot with the log verbosity cranked up, I then found this transaction was erroring out in `master`:
```
[... DEBUG solana_runtime::message_processor] Program eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU invoke [1]
[... DEBUG solana_runtime::message_processor] Program eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU consumed 226 of 200000 compute units
[... DEBUG solana_runtime::message_processor] Program failed to complete: Access violation in heap section at address 0x30003ff98 of size 8 by instruction #148285
[... DEBUG solana_runtime::message_processor] Program eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU failed: Program failed to complete
[... DEBUG solana_runtime::bank] check: 5us load: 2101us execute: 82893us txs_len=1
[... DEBUG solana_runtime::bank] tx error: Err(InstructionError(0, ProgramFailedToComplete)) Transaction { ... }
```

I have made the ledger available [here](https://drive.google.com/file/d/1QbRe43r9G8s-o-3ldkD6qrd2JRschMil/view?usp=sharing).
I have been starting the validator with [this script](https://github.com/solana-labs/solana/files/7508187/val_run.txt); adjust paths for log/ledger/keypair arguments as well as executable path to fit your setup.

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

In a recent re-org the compute budget bump checks got ignored

#### Summary of Changes

- add them back with tests

Fixes #21225

### CODE REVIEW

**Comment by steviez** [comment]
I'll defer to others for proper review, but running the tip of master (5fb0ab9d000944aee045fd5ea619e826f4b80a6f) with this commit on top of it, I no longer see the tx failure described in https://github.com/solana-labs/solana/issues/21225

**Comment by Lichtso** [comment]
https://github.com/solana-labs/solana/blob/cc3c593cfc18a7b743353ea31ed0ea58ffdc4518/program-runtime/src/instruction_processor.rs#L365

Can you change
`if let Ok(root_account) = keyed_account_at_index(keyed_accounts, 0) {`
to
`let root_account = keyed_account_at_index(keyed_accounts, 0)?;`

That changes the error from `InstructionError::UnsupportedProgramId` to `InstructionError::NotEnoughAccountKeys`.
But, because that should now be covered by the earlier check in `InvokeContext::push()`, it is not possible to fail here anymore.

**Comment by jackcmay** [comment]
> https://github.com/solana-labs/solana/blob/cc3c593cfc18a7b743353ea31ed0ea58ffdc4518/program-runtime/src/instruction_processor.rs#L365
> 
> Can you change `if let Ok(root_account) = keyed_account_at_index(keyed_accounts, 0) {` to `let root_account = keyed_account_at_index(keyed_accounts, 0)?;`
> 
> That changes the error from `InstructionError::UnsupportedProgramId` to `InstructionError::NotEnoughAccountKeys`. But, because that should now be covered by the earlier check in `InvokeContext::push()`, it is not possible to fail here anymore.

Why not leave it as is?

**Comment by Lichtso** [comment]
> Why not leave it as is?

To test if anything still depends on the error to be caught in `InstructionProcessor::process_instruction()` but not in `InvokeContext::push()`. And, to get rid of an extra "if"-nesting-level.

**Comment by jackcmay** [comment]
Yup, makes sense, unnecessary

File: program-runtime/src/invoke_context.rs [hunk 1]
```diff
@@ -103,6 +103,7 @@ pub struct ThisInvokeContext<'a> {
     sysvars: &'a [(Pubkey, Vec<u8>)],
     logger: Rc<RefCell<dyn Logger>>,
     compute_budget: ComputeBudget,
+    current_compute_budget: ComputeBudget,
```
Comment (Lichtso) [inline program-runtime/src/invoke_context.rs:106]: What is `current_compute_budget` and how is it different from `compute_budget` ?

**Review by Lichtso — APPROVED** [review]
One more question, otherwise looking good.

File: program-runtime/src/invoke_context.rs (hunk 1 above)
Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:106]: Since the bump is only for a single instruction `compute_budget` which is the original is used to reset `current_compute_budget` for additional instructions that don't meet the bump criteria.

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:106]: If no further concerns I'll commit it

File: program-runtime/src/invoke_context.rs [hunk 2]
```diff
@@ -210,11 +212,32 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
                 // Reentrancy not allowed unless caller is calling itself
                 return Err(InstructionError::ReentrancyNotAllowed);
             }
-        }
+            *program_id
+        } else {
+            return Err(InstructionError::UnsupportedProgramId);
```
Comment (jstarry) [inline program-runtime/src/invoke_context.rs:217]: @jackcmay @Lichtso it looks like this new err has made master and v1.8 incompatible. On v1.8 it's possible to invoke the native loader directly from a tx in which case `program_indices` is empty and the tx can still be processed successfully.

Here is an example tx that showcases the different behavior. The tx invokes the native loader in a way that runs the code path for a 0 lamport system ix transfer with the program_id set to the native loader id: https://explorer.solana.com/tx/inspector?message=AQABAxqaPUT28ZZYeGqKkcOfzL7aLBsifPzASKGnxP4QwCbm%252Bk%252BRkBm4gzLzoCATvtlllmIDwh%252B8XErqP9gYsdFvpjUFh4S%252FFIukKC%252BwEldIiKnxU6B9rfdlwEVcmpcDgAAAAL1QHO0HDnEf1DbxkmiM2jg37QuLnqbQDF9g5Y7HZgN3AQICAAEMAgAAAAAAAAAAAAAA

It was constructed with the following JS:
```js
    // create this to copy out the ix data    
    const ix = SystemProgram.transfer({
        fromPubkey: MY_KEY.publicKey,
        toPubkey: NEW_KEY.publicKey,
        lamports: 0,
    });
    
    const tx = new Transaction().add(new TransactionInstruction({
        programId: new PublicKey("NativeLoader1111111111111111111111111111111"),
        data: ix.data,
        keys: [{
            pubkey: MY_KEY.publicKey,
            isSigner: false,
            isWritable: false,
        }, {
            pubkey: NEW_KEY.publicKey,
            isSigner: false,
            isWritable: true,
        }],
    }));
```

Comment (Lichtso) [reply program-runtime/src/invoke_context.rs:217]: As mentioned [here](https://github.com/solana-labs/solana/pull/20448/files#r752310865) I think we need to either add an empty transaction to the test corpus or feature gate and forbid them. Because they are breaking the assumption that there is at least one executable account loaded which can be the `program_id`.

Comment (jstarry) [reply program-runtime/src/invoke_context.rs:217]: I'm in favor of feature gating and forbidding them. But we will still need to revert behavior and add a test case for this in the meantime until we enable that feature switch.

Comment (Lichtso) [reply program-runtime/src/invoke_context.rs:217]: Agreed in both, the temporary test and the plan to remove them.
It also makes life a little easier in terms of edge cases which we need to handle.

To do so:
- remove the line `debug_assert_eq!(first_instruction_account, 1);` in `process_instruction_with_program_logging()`
- add `.map_err(|_| InstructionError::UnsupportedProgramId)` behind `keyed_account_at_index(keyed_accounts, 0)?` in `instruction_processor::process_instruction()`.
- pull `InstructionError::ReentrancyNotAllowed` check out of `let program_id = if let Some(index_of_program_id) = program_indices.last() {` in `ThisInvokeContext::push()`
- and then put the `let program_id = if let Some(index_of_program_id) = program_indices.last() {` in `if self.invoke_stack.is_empty() {` without an else return error branch.
- finally, add the empty transaction test case in `program-runtime/src/invoke_context.rs`

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:217]: The change that are causing this behavior were an optimization to get the program_id in push.  We don't have to do that and instead allow empty indices and get the program id from the message instead.

Comment (Lichtso) [reply program-runtime/src/invoke_context.rs:217]: We are tracking this in a separate issue: #21346
